### ISSUE TITLE
Large portions of the window not repainting on high DPI/resolution displays

### ISSUE THREAD

**Issue by BytewaveMLP** [Microsoft/vscode#45492]
Issue Type: <b>Bug</b>

When VS Code sits open for a while, say in the background or after an unmeasured time of just sitting around, large portions of the window appear to not repaint. This also occurs with more "complicated" views, such as the terminal or git diff views, upon scrolling. This causes an insane amount of flickering, and in git diff views it's almost impossible to use as large portions of the screen don't repaint upon scrolling.

This does **not** occur in other Electron apps, such as Discord or GPMDP, nor does it appear in Chromium or Chrome (not that I suppose that matters).

I know this is a bit of a duplicate of old issues, as I've seen issues like #12473 before, but all of those have been marked as resolved. I don't know if it's my system specifically or what, but none of the fixes in any of these issues seem to work; not even `--disable-gpu`, `--force-gpu-rasterization`, `window.zoomLevel: 0`, or any of the other tricks have had any effect.

I don't see it mentioned in the system info, so I suppose I should also note that I'm running on an AMD RX 480 with `mesa` and the AMDGPU open source drivers, on KDE X11.

### Screenshots

(Both of these are more related to the git diff issue, but I don't have a good way to record nor reproduce the time-based flickering issue yet.)

![image](https://user-images.githubusercontent.com/5623770/37247690-c93f8a34-2484-11e8-88c1-fc7ebdcd94f1.png)

> Launched with `--disable-extensions`.  Notice how a lot of the code is missing, and the terminal has a very large section of gray.

![image](https://user-images.githubusercontent.com/5623770/37258779-bdf0834c-254a-11e8-892e-a239260c027b.png)

> Yet another screenshot, this one with more pronounced differences. Interestingly, scrolling around in the git diff view seems to partially fix the terminal, and vice versa, even though the panel being scrolled becomes more broken.

### Steps to reproduce

1. Launch VS Code (even with `--disable-gpu` or `--force-gpu-rasterization`)
2. Open a git diff view, or let the editor sit for a while
3. Observe obnoxious rectangles

VS Code version: Code 1.21.0 (9a199d77c82fcb82f39c68bb33c614af01c111ba, 2018-03-07T11:01:43.521Z)
OS version: Linux x64 4.15.7-1-zen

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|AMD FX(tm)-8350 Eight-Core Processor (8 x 3734)|
|Load (avg)|1, 2, 1|
|Memory (System)|15.61GB (5.59GB free)|
|Process Argv|/opt/visual-studio-code/code|
|Screen Reader|no|
|VM|0%|

</details><details><summary>Extensions (51)</summary>

Extension|Author (truncated)|Version
---|---|---
language-x86-64-assembly|13x|2.2.11
vscode-markdownlint|Dav|0.13.0
EditorConfig|Edi|0.12.1
vsc-material-theme|Equ|1.5.1
much-assembly-required-language-support|PJB|0.1.5
vscode-docker|Pet|0.0.25
code-settings-sync|Sha|2.9.0
project-manager|ale|0.24.1
vscode-twig-pack|baj|1.0.0
love|bsc|0.3.6
npm-intellisense|chr|1.3.0
path-intellisense|chr|1.4.2
gitignore|cod|0.5.0
vscode-eslint|dba|1.4.7
git-extension-pack|don|0.1.3
githistory|don|0.4.0
gitlens|eam|8.1.0
tslint|eg2|1.0.28
vscode-npm-script|eg2|0.3.3
prettier-vscode|esb|1.2.2
php-debug|fel|1.12.1
php-intellisense|fel|2.2.9
auto-close-tag|for|0.5.6
code-runner|for|0.9.1
discord-vscode|icr|2.2.1
search-node-modules|jas|1.3.0
csharpextensions|jch|1.3.0
docthis|joe|0.6.0
docomment|k--|0.0.17
Go|luk|0.6.77
MagicPython|mag|1.0.12
license-injector|mar|0.0.2
dotenv|mik|1.0.1
prettify-json|moh|0.0.3
vscode-apache|mrm|1.1.1
python|ms-|2018.2.1
cpptools|ms-|0.15.0
csharp|ms-|1.14.0
PowerShell|ms-|1.6.0
debugger-for-chrome|msj|4.2.0
php-docblocker|nei|1.3.3
proto|pet|0.0.2
livescript|pse|0.1.1
sass-indented|rob|1.4.8
code-spell-checker|str|1.6.5
much-assembly-required-upload-on-save|tom|0.0.6
gitblame|wad|2.2.0
nodejs-extension-pack|wad|0.1.9
vscode-import-cost|wix|2.6.2
JavaScriptSnippets|xab|1.5.0
vscode-open-in-github|ziy|1.3.1

</details>
Reproduces without extensions

**Comment by bpasero** [comment]
@BytewaveMLP does it reproduce when you run with `code --disable-gpu`?

**Comment by BytewaveMLP** [comment]
@bpasero Yes. Both `--disable-gpu` and `--disable-gpu-rasterization` had no effect. Extensions were disabled while testing.

**Comment by BytewaveMLP** [comment]
I feel like this has something to do with higher DPI displays. I've noticed that running Code in a smaller window seems to lessen if not eliminate the effect (but it also makes it much less usable).

For reference, this is a 4k display with 1.7 DPI scaling in KDE. But, again, other Electron applications running at the same resolution are unaffected (Discord used to be, but that was months ago; perhaps it's some sort of regression?).

My apologies for not mentioning this earlier. It hadn't really crossed my mind.

**Comment by danloiterton** [comment]
I have the same issue, running ubuntu 16.04 on a 4k monitor with an AMD graphics card (`AMD OLAND (DRM 2.43.0 / 4.4.0-116-generic, LLVM 5.0.0)`). 

It seems to have worsened since the terminal splitting feature (which is so awesome to use on my windows laptop, but unusable on my ubuntu setup). The git diff spilt view has been causing flickering and blank rectangles for some time but I got by without using it. Now, however, the flickering is happening even when there's no screen splitting enabled. 

I've also tried running code with  `--disable-gpu` and `--force-gpu-rasterization`. This seems to improve the git split screen a little but certainly doesn't fix the problem, and the terminal splitting remains unusable.

Similar to @BytewaveMLP, if I reduce the window size, the problem is alleviated somewhat, but that does defeat the purpose of my shiny 4k monitor!

![image](https://user-images.githubusercontent.com/30551212/37590883-86f4d8d0-2b61-11e8-8140-4ab46418fe4a.png)

notice big white, grey and dark grey rectangles over the side panel and terminal windows in the example above. But it happens all over the screen, usually triggered by scrolling. Often there's flickering but it's when they don't go away and the code is obscured, that the app becomes unusable

**Comment by mix3d** [comment]
I've experienced the issue on Windows 10, laptop has a native WQHD 1440p screen, set to 150% scaling, but mostly use a larger monitor at 100% scaling. I also get strange painting issues with terminals, and text sometimes disappears altogether until I type some commands, or the position of the current active line gets moved INTO the existing displayed console output.

**Comment by Tyriar** [comment]
This might workaround the problem:

```
code --ignore-gpu-blacklist
```

**Comment by mix3d** [comment]
@tyriar how do you set code to always launch with a command?

**Comment by Tyriar** [comment]
@mix3d did it work? On Windows I think you could create a shortcut and modify it, or launch from a batch script.

**Comment by BytewaveMLP** [comment]
Yup, it seems like `--ignore-gpu-blacklist` fixes it @Tyriar; however, it's annoying to have to edit all my `.desktop` files to support this. Fortunately, KDE makes that easy, but it's kinda difficult to set `$EDITOR` with flags (Code doesn't seem to like the filename being passed *after* command line arguments, but perhaps that's another bug to report).

**Comment by mix3d** [comment]
Sounds like a good opportunity to open that value as an editor config? Or is that too far into the app loading process?

**Comment by BytewaveMLP** [comment]
It could at least be a prelaunch thing. On Arch, the Google Chrome package "binary" actually Chrome from a shell script that reads a file `~/.config/chrome-flags.conf` and uses its contents as command line arguments to pass to the actual Chrome instance. Something like that could be used here.

**Comment by Tyriar** [comment]
@BytewaveMLP great, we have a root cause! This is happening because Chromium maintains a list of GPUs to disable as they have issues, unfortunately in VS Code this manifests itself as corrupted textures inside the canvas for some reason.

I'm curious if the same thing happens when you run Google Chrome?

> Sounds like a good opportunity to open that value as an editor config? Or is that too far into the app loading process?

@mix3d I think it's too far, we don't touch the settings file until the Electron main process is launched.

**Comment by BytewaveMLP** [comment]
@Tyriar It's hard to say. Chrome doesn't have any flickering issues when the `ignore-gpu-blacklist` flag is disabled, but it does seem considerably less performant, and changing views (e.g. switching tabs) does seem to cause a sort of "wiping" effect much like VS Code was exhibiting. Discord, another Electron app, used to have a similar problem to the one noted in this issue, but it would only appear at certain startups and could be fixed after a client or system restart; it was never reliable reproducible, and was fixed a while ago )likely ignoring the blacklist themselves). [GPMDP](https://github.com/MarshallOfSound/Google-Play-Music-Desktop-Player-UNOFFICIAL-), however, does have a slight flickering and poor performance problem, which does seem related.

**Comment by Tyriar** [comment]
@BytewaveMLP oh it should only show up when you're using a canvas though. To test this out you could run the xterm.js demo (which is what the terminal is built on) as that's a canvas that's doing a lot of painting. Try tweaking the rows/cols in the textbox below and running some commands. https://github.com/xtermjs/xterm.js#linux-or-macos

**Comment by BytewaveMLP** [comment]
Er... whoops. I guess I missed the canvas part initially. Sorry about that.

However, even with the GPU blacklist disabled and something that should cause a lot of repainting `hd /dev/urandom`, I'm unable to get any flickering out of Chrome, and only some poor performance, even with an exceptionally large terminal size (500x100, anyone?).

**Comment by BytewaveMLP** [comment]
@danloiterton Don't forget that Chrome has `chrome://flags` which should let you set the `--ignore-gpu-blacklist` flag without having to pass it on the command line. Code doesn't have a similar interface, unfortunately, but at least you don't have to make a shell script for *everything*. :P

**Comment by jeremyfa** [comment]
Same issue here, started to have it after updating vscode last week I think. I have it on a Macbook Pro Retina 15 inches, but it doesn't occur on my secondary screen which is default DPI.

**Comment by mix3d** [comment]
what's the impact if we turn it on for all builds as part of the regular
release?

**Comment by BytewaveMLP** [comment]
I believe the list of disabled/enabled cards is available [here](https://www.khronos.org/webgl/wiki/BlacklistsAndWhitelists#Chrome). Any systems with these cards disabled with good reason may be affected, but considering the age of most of these... hm.

**Comment by Tyriar** [comment]
@BytewaveMLP it looks like [this file](https://cs.chromium.org/chromium/src/gpu/config/software_rendering_list.json?q=software_rendering_list&sq=package:chromium&dr) is what's used and goes into detail on what features are disabled for each configuration. `accelerated_2d_canvas` is the primary feature we need for this, not 100% sure what other ones affect canvas rendering.

**Comment by Tyriar** [comment]
@mix3d maybe it's this one? https://cs.chromium.org/chromium/src/gpu/config/software_rendering_list.json?q=software_rendering_list&sq=package:chromium&dr&l=748

You could try updating your drivers. Intel is mentioned a lot in that file, it's tough to tell without knowing. Also we'll be on an older version of that file since Electron isn't on the latest version of Chromium, so the list will be a little different for us.

**Comment by martinmeinke** [comment]
I've got the same issue on a Lenovo P50 (4K display) in an Ubuntu 14.04 VM. The issue to get worse with window size. --ignore-gpu-blacklist does not seem to resolve the issue.

**Comment by Tyriar** [comment]
@martinmeinke GPU support always seems to be broken in VMs, try --disable-gpu and also disabling GPU acceleration in the VM settings if available.

**Comment by BytewaveMLP** [comment]
No, bad bot. We had something going here. 😡

**Comment by teki** [comment]
I think this was never actually fixed, there was a workaround in Chrome: https://bugs.chromium.org/p/chromium/issues/detail?id=695427
But that did not actually fix the issue.
I got used to just restart VSCode, but it just gets broken sooner and sooner with new releases (the terminal seem to be affecting it badly).
1.22.1 made the problem worse again.
(my machine: MBPro 15inch, mid2015, macOs 10.13.4, 16GB ram, Intel Iris Pro 1536 MB ram)
The only really working "workaround" right now is to run the external display in native resolution without scaling.

**Comment by methodbox** [comment]
This was bugging me so I did a little digging. Found a nice list of command line switches for Chromium here: https://peter.sh/experiments/chromium-command-line-switches/.

When I launch VS Code from Terminal with $code --disable-gpu-compositing the issue is resolved.

Note that during my search I also found that there seems to be a known issue relating to the cursor blink - I messed with this as the flicker seemed to happen when I changed the focus - swapping tabs, moving the cursor - which would also change the cursor focus. This didn't seem to have an effect, but I found using `"editor.cursorBlinking": "smooth"` drops the CPU usage down from about 20% to 8%, and sometimes it's nearly idle with VS Code open.

If you're a perfectionist (like me) and you want to have a standard "app" to open like this (as you may not always be in terminal) you can use this simple AppleScript in Automator:

`
on run
	do shell script "/usr/local/bin/code --disable-gpu-compositing"
end run
`
Save as an application (yourapp.app).

If you want to get even pickier (like me) you can find the VS Code app in your Applications folder -> Get Info and click the icon at the top left -> Cmd+C then do the same on yourapp.app and Cmd+D and now it loosk like VS Code proper.

Hope that helps until they fix this.

**Comment by jlmaners** [comment]
Had previously experienced the flickering but it resolved itself with an update. Now the flickering is back. 

OS: macOS 10.13.4
VSCode: 1.22.2

All while using HiDPI display connected to my mid-2015 15" MBP with AMD Radeon R9 M370X graphics card. As @teki mentioned, only work around that I've had success with is running in native resolution.

**Comment by methodbox** [comment]
@jlmaners
I don’t believe this is likely GPU-specific. I’m on a 2016 MBP with an AMD GPU but others with Nvidia have experienced the same issue.

It’s an issue that is essentially the same as one that plagued Atom (also built in Electron).

I would take a look at my post and the link to Chrome switches - there’s many related to GPU.

If one or the other doesn’t do the trick (or multiple together) I’d be very surprised.

If this were a driver issue there would be more pixel drawing on the screen that is affected than just VS Code (like the OS background) - you’ll notice in all examples there isn’t.

This means the rendering/drawing latency is isolated to the VS Code instance and has something to do with how it’s talking to the GPU (which is why I started with a buffer setting).

**Comment by trapgate** [comment]
I've run into this on a mac as well. None of the gpu flags I tried made any difference, but the window size did - anytime the window was over a certain size the problem showed up. That size was about 2/3 of the monitor.

The problem only shows up for me when the view is split for viewing files side-by-side. Disabling scaling on the monitor made the problem disappear completely. This monitor is 3840x2160. The scaling option is set to, "looks like 3008x1692". It seems to be total window area that triggers the problem, not just a window taller or wider than some limit.

**Comment by teki** [comment]
The original chrome bug is about running out of GPU memory: https://bugs.chromium.org/p/chromium/issues/detail?id=695427

**Comment by methodbox** [comment]
Changing the buffer size (or disabling it) should free up real time available memory, so maybe that’s why it worked for me but I’d also expect there to be some serious fan action if the 4GB of GPU memory were maxed - not the case.

There are also a number of flags related to CPU and software rendering - can you guys play with these and see what happens?

**Comment by methodbox** [comment]
Might want to look at —device-scale-factor.

Since all of these potentially set a value which also affects processes down the line it’s likely one of those.

There’s a number of raterization flags re: CPU vs. GPU.

**Comment by methodbox** [comment]
"editor.disableTranslate3d": true

https://github.com/Microsoft/vscode/issues/25934

Grasping at straws here, but safe to say this is something that is fixed in Electron and Chrome for a while but somehow that fix is not backported into VS Code.

**Comment by teki** [comment]
I already had `"editor.disableTranslate3d": true`.
`--high-dpi-support=1 --force-device-scale-factor=1` might work, but it is not something that you want to look at.
`--high-dpi-support=1` alone works and seems to be an acceptable workaround, it slows the UI down, but bearable.

**Comment by methodbox** [comment]
It may slow the UI down, but that’s a boolean that should probably be part of VS Code settings.

It’s enabled for Atom by default without noticeable UI issues.

I’m really starting to think that this is a VS Code-specific issue with their Electron config.

**Comment by drewlustro** [comment]
Please – bumping for attention on this issue. I will try `--hi-dpi-support=1` today, but I am near certain this is a VSCode issue as other Electron apps do not suffer in the same manner.

**Comment by vgeddes** [comment]
On my setup, with vscode 1.23 (Insiders edition), passing `--disable-gpu-blacklist` fixed the issue once and for all. 

Platform: Linux
OS: Ubuntu 16.04
GPU: Radeon RX 550 2GiB

**Comment by vgeddes** [comment]
FYI There are also bleeding edge "exploration" builds available  - these use Electron 2.0 and a newer Chrome: https://github.com/Microsoft/vscode/issues/45542

I'm now using the exploration build together with `--disable-gpu-blacklist` for a flicker and artifact free experience.

**Comment by methodbox** [comment]
No, this is definitely a VS Code issue, I think @vgeddes and others' testing shows that. Now we just need to probably wait for the next build that they backport the Electron fixes.

**Comment by masonk** [comment]
It's back since the latest update,

**Comment by methodbox** [comment]
I just tested as I was having a syntax issue which turned out to be unrelated, but this issue is still occurring.

This is my solution for now (from my comment on Apr 9 above):

> If you're a perfectionist (like me) and you want to have a standard "app" to open like this (as you may not always be in terminal) you can use this simple AppleScript in Automator:

on run do shell script "/usr/local/bin/code --disable-gpu-memory-buffer-compositor-resources" end run
Save as an application (yourapp.app).

If you want to get even pickier (like me) you can find the VS Code app in your Applications folder -> Get Info and click the icon at the top left -> Cmd+C then do the same on yourapp.app and Cmd+D and now it loosk like VS Code proper.

Hope that helps until they fix this.

**Comment by wolfhoundjesse** [comment]
I'm trying the `--disable-gpu-memory-=buffer-compositor-resources` advice now.

My issue is slightly reversed. I have a shiny new Dell Precision 5520 with a 4k display, and I'm using two older external monitors that aren't. Flickering and randomly crashing are par for the course. Most frustrating is when my integrated terminal window doesn't repaint, and I have to drag the size up or down to get it back. Eventually anything I mouse over flickers, and I have to restart. Also, I have to turn off the laptop monitor, or system resources go haywire while Code is open. That's probably another issue entirely.

I've ordered a pair of 43" 4K monitors from LG. My thinking is that if I remove scaling, this issue will go away. It works perfectly without the external monitors. After reading this newest set of issues, I'm not sure that will solve the problem.

Arch Linux: 4.14.51-1-lts
i3wm

**Comment by methodbox** [comment]
@wolfhoundjesse

Unless you’re using a macOS GPU driver and API in Arch, I’d suspect not.

There are a number of other GPU flags that might help. I provided a link much earlier in the thread.

**Comment by mix3d** [comment]
I experience this problem on Windows, so I don't think the drivers are that
important.

**Comment by methodbox** [comment]
@mix3d 
Are you also using an AMD GPU? My MBP has the AMD 550. It seems most everyone in this thread is using AMD.

There must be something in common.

**Comment by mix3d** [comment]
Intel HD 620. Lenovo X1 yoga laptop.

**Comment by methodbox** [comment]
Ran into this with the latest version. Seems that ```--high-dpi-support=1 ``` now seems to solve it for me, which is weird because it didn't previously.

**Comment by teki** [comment]
I have the opposite experience with `--high-dpi-support=1`, it used to work, now the glitches are back with 1.24.1. `--disable-gpu` does not do anything either.

macOS 10.13.5 (MacBook Pro 15 mid 2015), Intel Iris Pro GPU, 4k external display.

**Comment by teki** [comment]
Managed to get it into usable shape with enabling reduce motion: https://support.apple.com/kb/PH25204
My guess is that Chrome is still running out of GPU memory.

**Comment by DuncanFaulkner** [comment]
I see something similar but just when in the terminal view (see below), using insiders version 1.25 on ubuntu 18.04. 
the terminal window, is sitting over the explorer, I have the exact same version on a 16.04 and this isn't an issue, I guess it may have something to do with monitor size (18.04 on 27" screens 16.04 on 22" screens) neither screens are 4k (as far I can tell).

![vs screen shot](https://user-images.githubusercontent.com/1952116/42346994-4e3c3aac-809c-11e8-9b3e-5fa83d31a68e.png)

**Comment by Tyriar** [comment]
@DuncanFaulkner I think that's https://github.com/Microsoft/vscode/issues/51479

**Comment by DuncanFaulkner** [comment]
@Tyriar OK thanks will take a look

**Comment by methodbox** [comment]
@ashmckenzie DisplayPort has significantly higher bandwidth. If you’re using anything 2K or higher for your resolution and want 60Hz or better refresh rate (and you do ;)) then yeah, you want DP.

I on the other hand have been using a LG USB-C monitor at 4K 60Hz. DisplayPort is run over USB-C in this case.

I can try my USB-C to

**Comment by agg23** [comment]
Similar to @BytewaveMLP, I am using DisplayPort exclusively. Beyond VSCode, I also experience similar flickering in Discord desktop, though I am running in Arch with a RX580, whereas it seems that most users here are using Macs. Definitely appears to be an Electron issue.

**Comment by methodbox** [comment]
@bpasero Thanks! I will try this later today and report back.

**Comment by Ts-OV** [comment]
Sadly confirm the similar issues. Unfortunately, reproducible for both UHD and HD monitors.
I have encountered this myself 1st time after updating VC from ver. 1.25 to recent 1.27 on Ubuntu 17.01 (run over XRDP from Xeon Server on HP laptop). However, one of my colleagues was reporting this for VC versions earlier than 1.25. In the latter case this was also U17 but run in VBox on W7 HP laptop.
In my case, 1st of all I tried reinstalling VC packages. This allowed to forget about the issue for about 1 week. After the issues got back, neither the packages reinstall, nor updates (nor re-copy of RDP fixing packages [https://github.com/Microsoft/vscode/issues/3451]) helped.
On the other hand, I noticed that the VCode-Insiders didn't have that issue (at least at the moment). I guessed this is caused (in addition to visual perception of what was happening on the screen, where some strange background color rectangular blocks started to appear randomly overlapping (covering, cleaning) the code window text) by some bug with mouse hover hints pop-ups. I increased the code hints popup time and then switched that off completely (although, didn't find how to do the same for full file names pop-ups that appear when mouseovering on top of the file-names shown e.g. in the file list of file browser area or on the top bar of open files tabs). The luck of the VCode-Insiders in this sense seemed to be in the inability to find proper inclusions, so VCode-Insiders wasn't able to show any hint for the code, and hence, no mouseover popups.
Sadly, but disabling the hint popups didn't help the VC. After sometime (even without any mouse moves or visible origin for doing mouse hover triggered actions), the same random background color overlapping rectangles started emerging randomly again, so messing up all the text in the code window (up to 80%). 
It doesn't look useful to provide any static screenshots. The videos of what happens might be more useful, if there are no guess from VC developers where the issue night come from..
Just let me know, whether it's needed then. I can shoot the video and post here. 
So far, crossing my fingers with hope the VCode-Insiders could work. Otherwise, i'm completely blocked with VC usage for my project development..

**Comment by Ts-OV** [comment]
Ok. After sometime (probably allowing the corresponding spare processes spoiling text window to finalize), deprecation of mouse over popups really helped. 
Next day I tried VC, there was just some VC "nervous" attempts to redraw with background color several lines below one was editing at the moment, but then it was able to recover the text back quite fast.
Only once it showed up with the origin of what probably was casing that. 
Seems to be there were VC attempts to show some strange typed code context helper popup. But the info on the popup helper was completely illogical and inconsistent! E.g. it was popping up after typing colon symbol in the double slash comment line. Why the heck?
![unknown_illogical_popups](https://user-images.githubusercontent.com/43061088/45253198-1a119580-b36c-11e8-8215-183a9dea40ff.jpg)
Is it a bug?

**Comment by Ts-OV** [comment]
That's how it looks like with no popups blocked:
![vcode_graphics_bugs_c](https://user-images.githubusercontent.com/43061088/45255603-615e4d00-b391-11e8-9ca2-b2d3a20a606e.jpg)

**Comment by Ts-OV** [comment]
Also, as the bug description implies, switching VC from full screen mode to windowed of smaller size on UHD monitors seems like allows decreasing the number of residual aftereffects of the mentioned graphics redraw issues.

**Comment by bpasero** [comment]
We are building exploration builds that use a much newer version of Electron (our UI framework). I wonder if this issue reproduces with one of these builds, could you try? Download:
* Windows: [Download](https://az764295.vo.msecnd.net/exploration/883208e30da25777874a8ade5f90e1fea638a298/VSCode-win32-x64-1.28.0-exploration.zip)
* macOS: [Download](https://az764295.vo.msecnd.net/exploration/883208e30da25777874a8ade5f90e1fea638a298/VSCode-darwin-exploration.zip)
* Linux: [Download](https://az764295.vo.msecnd.net/exploration/883208e30da25777874a8ade5f90e1fea638a298/code-exploration-1535724826.tar.gz)

**Comment by Ts-OV** [comment]
Ok. After waiting for MS IntelliSense to setup (in order to be able showing mouseover popup hints), it looks like I do not see the graphics issues (Linux U17, XRDP) in the corresponding package after playing a little.

**Comment by Ts-OV** [comment]
What I meant previously, is VC standard extension (offered automatically by VC to be installed based on C/C++ project I work currently): 
"
C/C++ | ms-vscode.cpptools | Microsoft
C/C++ IntelliSense, debugging, and code browsing.
"
In addition to standard installation, restarting, etc. more or less quick done actions, it need time to build up its database in order to find the proper sources (includes) to show hints for function declared either by mouseover or when intentionally clicked (from right-click menu) "Go to Definition/Declaration".
Until this DB is built, the hints cannot be shown, so there is no so heavy graphics redraw load (no mouseover based popups).

**Comment by rightaway** [comment]
Is anyone else seeing this issue only in the git file diff view? When a file has many changes, vscode becomes almost unusable until I close that diff tab, then things go back to normal. If the file has just a couple of changes it's not a problem. So I need to use git diff in the terminal not in vscode.

I don't have a non-HIDPI monitor to test against so I assume it's related to this issue.

**Comment by limptwiglet** [comment]
@rightaway I'm still seeing this on Ubuntu with i3wm.
It seems to be caused when ever there are splits showing (which the diff has). One way around this (for me at least) is to pass the --disable-gpu-blacklist flag when I start code

**Comment by jovere** [comment]
@bpasero 
I have had the same issue with just a 1920x1080 monitor in ubuntu.  The option `C_Cpp.dimInactiveRegions` turns the entire screen to empty blocks if any large sections of code are dimmed.  It looks very similar to the picture from @Ts-OV and even ends up misdrawing the left side panel.  I also needed to set the terminal renderer to `dom` in order for it to be responsive.

However, with the exploration version that was posted a couple weeks ago, those issues appear to be gone.  There aren't any empty blocks and the terminal updates very quickly with the default renderer.

**Comment by kobalicek** [comment]
Affected as well, but this happened after updating nvidia driver (Linux). Not sure what is the problem, but it means that VSCode is unusable to me. I tried `--disable-gpu` parameter, but with no luck.

**Comment by kobalicek** [comment]
I was only lucky with `--disable-gpu --force-cpu-draw` arguments. It seems that `--disable-gpu` is not enough. This is definitely Blink or SKIA issue, VSCode is just a victim in this case :)

**Comment by rightaway** [comment]
I had to remove `--disable-gpu-blacklist` flag because it was causing very high CPU usage since very recently, possibly the last monthly release. But for some reason I'm not getting repainting issue anymore in the diff view, so the problems seems to be resolved for me at least for now. Anyone else?

**Comment by mix3d** [comment]
Latest build seems to be working well for me too. Have not noticed paint issues in a while. I tried the new TypedArray for the console, but had a crash or two so I disabled it.

**Comment by abtrumpet** [comment]
Ok, thanks. I will try the `electron-3.0.x` branch and see how it goes. It generally manages to run better, but I was having issues with it occasionally freezing. Such is the life of  HiDPI laptop owner :)

**Comment by bpasero** [comment]
Closing this issue given that we plan to release VSCode stable early February with Electron 3.x. If you want to benefit from the fix already, consider to use our insiders version that already contains the fix: https://code.visualstudio.com/insiders/

**Comment by cscomfort** [comment]
As of this writing, I'm using the latest version of the Insiders build on MacBook Pro (15-inch, Radeon Pro 560X 4096 MB) + OS X 10.14.1 + Dell P2415Q (DisplayPort / MST Off / 3840x2160 @ 60Hz) and find that the issue still occurs.

After some additional troubleshooting, I determined that **in my case the cause seems to be related to adjusting the Resolution to Scaled in the OS X Display Preference pane**. With scaling disabled, I don't encounter the issue.

Note that using a third-party display resolution tool, [SwitchResX](http://www.madrau.com), seems to allow me to adjust resolution while at the same time avoiding the VSC flicker.

This may prove applicable for some.

**Comment by teki** [comment]
@cscomfort I have similar issues, reported as a new issue: https://github.com/Microsoft/vscode/issues/68188
